File: backend/garminview/api/schemas/actalog.py
from __future__ import annotations
from datetime import datetime
from pydantic import BaseModel


class WorkoutListItem(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    workout_date: datetime | None
    workout_name: str | None
    workout_type: str | None
    total_time_s: int | None


class MovementItem(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    workout_id: int | None
    movement_id: int | None
    sets: int | None
    reps: int | None
    weight_kg: float | None
    time_s: int | None
    distance_m: float | None
    rpe: int | None
    is_pr: bool
    order_index: int | None


class WodItem(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    workout_id: int | None
    wod_id: int | None
    score_value: str | None
    time_s: int | None
    rounds: int | None
    reps: int | None
    weight_kg: float | None
    rpe: int | None
    is_pr: bool
    order_index: int | None


class WorkoutDetail(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    workout_date: datetime | None
    workout_name: str | None
    workout_type: str | None
    total_time_s: int | None
    notes: str | None
    movements: list[MovementItem]
    wods: list[WodItem]


class SessionVitals(BaseModel):
    workout: WorkoutDetail
    has_vitals: bool
    hr_series: list[dict] = []
    body_battery: list[dict] = []
    stress: list[dict] = []


class MovementRef(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    name: str | None
    movement_type: str | None


class MovementHistoryItem(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    workout_id: int | None
    sets: int | None
    reps: int | None
    weight_kg: float | None
    time_s: int | None
    rpe: int | None
    is_pr: bool
    workout_date: datetime | None  # joined from workout


class PRItem(BaseModel):
    model_config = {"from_attributes": True}
    movement_id: int
    movement_name: str | None
    movement_type: str | None
    max_weight_kg: float | None
    max_reps: int | None
    best_time_s: int | None
    workout_date: datetime | None


class CrossRefItem(BaseModel):
    workout_date: datetime | None
    workout_name: str | None
    workout_type: str | None
    total_volume_kg: float | None
    body_battery_max: int | None
    hr_resting: int | None
    sleep_score: int | None
    stress_avg: int | None


class ActalogConfigOut(BaseModel):
    url: str | None
    email: str | None
    weight_unit: str | None
    sync_interval_hours: int | None
    sync_enabled: bool
    last_sync: str | None


class ActalogSyncStatus(BaseModel):
    last_sync: str | None
    status: str | None
    records_upserted: int | None
    error_message: str | None


class ActalogConfigIn(BaseModel):
    url: str | None = None
    email: str | None = None
    password: str | None = None
    weight_unit: str | None = None
    sync_interval_hours: int | None = None
    sync_enabled: bool | None = None


# ---------------------------------------------------------------------------
# Notes parser schemas
# ---------------------------------------------------------------------------

class NoteParseItem(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    workout_id: int | None
    workout_name: str | None = None
    workout_date: datetime | None = None
    content_class: str | None
    parse_status: str | None
    parsed_at: datetime | None
    reviewed_at: datetime | None
    error_message: str | None
    llm_model: str | None
    raw_notes: str | None
    formatted_markdown: str | None = None  # from parsed_json
    parsed_json: str | None
    parse_duration_s: float | None = None
    llm_tokens_prompt: int | None = None
    llm_tokens_generated: int | None = None
    llm_inference_s: float | None = None


class NoteParseQueue(BaseModel):
    total: int
    items: list[NoteParseItem]


class ParserRunResult(BaseModel):
    processed: int
    pending: int
    skipped: int
    errors: int


class ParserModelStats(BaseModel):
    model: str | None
    n: int
    avg_wall_s: float | None
    avg_inference_s: float | None
    avg_tokens_prompt: float | None
    avg_tokens_generated: float | None
    min_wall_s: float | None
    max_wall_s: float | None


class ParserStats(BaseModel):
    total: int
    by_status: dict[str, int]
    by_model: list[ParserModelStats]


class ParserConfigOut(BaseModel):
    ollama_url: str | None
    model: str | None
    min_note_length: int | None
    system_prompt: str | None


class ParserConfigIn(BaseModel):
    ollama_url: str | None = None
    model: str | None = None
    min_note_length: int | None = None
    system_prompt: str | None = None


class NoteParseApproveIn(BaseModel):
    formatted_markdown: str | None = None  # optional human edit before approve
    performance_notes: str | None = None


class ParserJobStatus(BaseModel):
    running: bool
    total_staged: int

File: frontend/src/stores/actalog.ts
import { defineStore } from "pinia"
import { ref, computed } from "vue"
import { api } from "@/api/client"

export interface WorkoutListItem {
  id: number
  workout_date: string | null
  workout_name: string | null
  workout_type: string | null
  total_time_s: number | null
}

export interface MovementItem {
  id: number
  workout_id: number | null
  movement_id: number | null
  sets: number | null
  reps: number | null
  weight_kg: number | null
  time_s: number | null
  rpe: number | null
  is_pr: boolean
  order_index: number | null
}

export interface WodItem {
  id: number
  workout_id: number | null
  wod_id: number | null
  score_value: string | null
  time_s: number | null
  rounds: number | null
  reps: number | null
  weight_kg: number | null
  rpe: number | null
  is_pr: boolean
  order_index: number | null
}

export interface WorkoutDetail extends WorkoutListItem {
  notes: string | null
  movements: MovementItem[]
  wods: WodItem[]
}

export interface SessionVitals {
  workout: WorkoutDetail
  has_vitals: boolean
  hr_series: { ts: string; hr: number }[]
  body_battery: { ts: string; value: number; type: string }[]
  stress: { ts: string; level: number }[]
}

export interface PRItem {
  movement_id: number
  movement_name: string | null
  movement_type: string | null
  max_weight_kg: number | null
  max_reps: number | null
  best_time_s: number | null
  workout_date: string | null
}

export interface CrossRefItem {
  workout_date: string | null
  workout_name: string | null
  workout_type: string | null
  total_volume_kg: number | null
  body_battery_max: number | null
  hr_resting: number | null
  sleep_score: number | null
  stress_avg: number | null
}

export interface ActalogConfig {
  url: string | null
  email: string | null
  has_password: boolean
  weight_unit: string | null
  sync_interval_hours: number | null
  sync_enabled: boolean
  last_sync: string | null
}

export const useActalogStore = defineStore("actalog", () => {
  const workouts = ref<WorkoutListItem[]>([])
  const selectedWorkout = ref<WorkoutDetail | null>(null)
  const sessionVitals = ref<SessionVitals | null>(null)
  const prs = ref<PRItem[]>([])
  const crossRef = ref<CrossRefItem[]>([])
  const config = ref<ActalogConfig | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)

  async function fetchWorkouts(start?: string, end?: string) {
    loading.value = true
    error.value = null
    try {
      const params: Record<string, string> = {}
      if (start) params.start = start
      if (end) params.end = end
      const { data } = await api.get("/actalog/workouts", { params })
      workouts.value = data
    } catch (e: any) {
      error.value = e.message
    } finally {
      loading.value = false
    }
  }

  async function fetchWorkoutDetail(id: number) {
    const { data } = await api.get(`/actalog/workouts/${id}`)
    selectedWorkout.value = data
  }

  async function fetchSessionVitals(id: number) {
    const { data } = await api.get(`/actalog/workouts/${id}/session-vitals`)
    sessionVitals.value = data
  }

  async function fetchPRs() {
    const { data } = await api.get("/actalog/prs")
    prs.value = data
  }

  async function fetchCrossRef(start?: string, end?: string) {
    const params: Record<string, string> = {}
    if (start) params.start = start
    if (end) params.end = end
    const { data } = await api.get("/actalog/cross-reference", { params })
    crossRef.value = data
  }

  async function fetchConfig() {
    const { data } = await api.get("/admin/actalog/config")
    config.value = data
  }

  async function saveConfig(updates: Partial<ActalogConfig> & { password?: string }) {
    // PUT /admin/actalog/config accepts a JSON body (not query params)
    await api.put("/admin/actalog/config", updates)
    await fetchConfig()
  }

  async function triggerSync() {
    const { data } = await api.post("/admin/actalog/sync")
    return data
  }

  const workoutsByDate = computed(() => {
    const map = new Map<string, WorkoutListItem[]>()
    for (const w of workouts.value) {
      if (!w.workout_date) continue
      const day = w.workout_date.slice(0, 10)
      const existing = map.get(day) ?? []
      existing.push(w)
      map.set(day, existing)
    }
    return map
  })

  return {
    workouts, selectedWorkout, sessionVitals, prs, crossRef, config,
    loading, error, workoutsByDate,
    fetchWorkouts, fetchWorkoutDetail, fetchSessionVitals,
    fetchPRs, fetchCrossRef, fetchConfig, saveConfig, triggerSync,
  }
})
